vi/day20.md

```diff
@@ -0,0 +1,484 @@
+
+`from()` also `completes` after it emits all values passed in.
+
+> Iterable is value that can be iterated over. For example: an Array, an Object, a Map, a Set, or a String. When you iterate over a String, you'll receive each character in that string.
+
+#### Array
+
+```ts
+/**
+ * output:
+ * - next: 1
+ * - next: 2
+ * - next: 3
+ * - complete
+ */
+from([1, 2, 3]).subscribe(observer);
+```
+
+When `from()` receives an `Array`, it emits the items in sequence similarly to `of(1, 2, 3)`.
+
+#### String
+
+```ts
+/**
+ * output:
+ * - next: 'h'
+ * - next: 'e'
+ * - next: 'l'
+ * - next: 'l'
+ * - next: 'o'
+ * - next: ' '
+ * - next: 'w'
+ * - next: 'o'
+ * - next: 'r'
+ * - next: 'l'
+ * - next: 'd'
+ * - complete: 'complete'
+ */
+from('hello world').subscribe(observer);
+```
+
+#### Map/Set
+
+```ts
+const map = new Map();
+map.set(1, 'hello');
+map.set(2, 'bye');
+
+/**
+ * output:
+ * - next: [1, 'hello']
+ * - next: [2, 'bye']
+ * - complete
+ */ 
+from(map).subscribe(observer);
+
+const set = new Set();
+set.add(1);
+set.add(2);
+
+/**
+ * output:
+ * - next: 1
+ * - next: 2
+ * - complete
+ */
+from(set).subscribe(observer);
+```
+
+#### Promise
+
+```ts
+/**
+ * ouput:
+ * - next: 'hello world'
+ * - complete
+ */ 
+from(Promise.resolve('hello world')).subscribe(observer);
+```
+
+In a case of `Promise`, `from()` will unwrap the `Promise` and `next` the `resolved` value (or `error` the `rejected`). This is also the _official_ way to convert a `Promise` to an `Observable.
+
+## `fromEvent()`
+
+`fromEvent()` is used to convert an `Event` to an `Observable`. For example, `DOM Event` like a mouse click or typing in an Input.
+
+```ts
+const btn = document.querySelector('#btn');
+const input = document.querySelector('#input');
+
+/**
+ * output:
+ * - next: MouseEvent { ... }
+ */
+fromEvent(btn, 'click').subscribe(observer);
+
+/**
+ * output:
+ * - next: KeyboardEvent { ... }
+ */
+fromEvent(input, 'keydown').subscribe(observer);
+```
+
+Notice that `fromEvent()` creates an `Observable` that does not `complete` after emission. This makes total sense because for events like `click` or `keydown`, we would want to keep listening to these events for as long as we need to. In other words, for as long as these DOM Elements are presented to the consumers. `fromEvent()` cannot determine on its own when we no longer need to listen to these events. This also means that we will need to `unsubscribe` from these `Observable` manually to avoid **memory-leak**.
+
+## `fromEventPattern()`
+
+`fromEventPattern()` is a _low-level_ version of `fromEvent()`. Conceptually, it is similar to `fromEvent()` that is to create `Observable` from events. However, the usage is quite different. `fromEventPattern()` is used to convert _unconventional_ or _callback-based_ Events to `Observable`. To understand this better, let's look at some example:
+
+```ts
+/**
+ * output: 
+ * - next: MouseEvent {...}
+ */
+fromEvent(btn, 'click').subscribe(observer);
```

Comment:
Here we say that `fromEventPattern` is used for unconventional events, but than bring up a conventional event (click) example that can be simply done via `fromEvent`. I think we need a better example. maybe something with WebSockets? I've seen an example of converting a WebSocket to Observable stream via `fromEventPattern`

Reply:
Ah, sorry, @nartc I didn't see that we have the WebSocket example up next. Maybe we should remove the click example, so that we don't overload the reader with unnecessary info?

Reply:
I've thought about this as well. Maybe remove `unconventional`? I want to have something that I can compare against `fromEvent()`. That's why I picked `click` from the `fromEvent()` example. What do you think? I really want to point out some APIs that have unconventional ways to listen for events.